Use dynamic linking for Tensorflow

apenugon opened the pull request:
@jailuthra 

RE https://github.com/livepeer/FFmpeg/issues/5

It turns out sharing the context wasn't enough - to get dlsym and dlopen to work I had to create some other pointer types. This code compiles but does not run though, so need to do some more debugging here for the MVP. This should give some context on the direction I'm going though.

apenugon committed:
Created a wrapper around the tensorflow library functions and replaced calls. This compiles, but fails to load sr model.

jailuthra commented:
Overall approach LGTM!

And sorry, I forgot to mention this - but our livepeer branch is quite old compared to ffmpeg master - and we're in progress to rebasing it on v4.4. So maybe use the 4.4 branch as the base https://github.com/livepeer/FFmpeg/tree/release/4.4

There are a ton of changes in the TF backend module that might lead to conflicts later - so better to avoid them now.

apenugon commented:
The update made just now will compile for a linux MVP that works with tensorflow, and does not force quit if run without tensorflow in LD_LIBRARY_PATH for non-tensorflow transcoding. 

I'm thinking I'll try rebasing on 4.4 last if that's OK, to make sure I get the Windows part done.

Speaking of that, anything I should watch out for when cross-compiling FFmpeg to windows in your experience?

apenugon commented:
@jailuthra This is tested and ready for review! Confirmed that this now links dynamically vs. statically with tensorflow, and tested the super-resolution filter on windows/linux with the env containing tensorflow and without.

Note that to build this you will need to still include tensorflow header files at build time (but not the lib).

jailuthra commented:
@apenugon great job! I just tested this with our scene classification filters on Linux with and without libtensorflow and it works :confetti_ball: 

I'll also setup a test env on windows and go through the code thoroughly sometime tonight - the first pass review looks all good to me.

One thing I noticed was that launching livepeer with scene classification filters enabled but without `libtensorflow` in the environment - it crashed with a segfault rather than a clean error exit. But if that's not reproducible in your tests with SR filter and ffmpeg tool directly (I didn't test that yet), then it might be how we initialize our filters - so no need to worry about it in this PR.

jailuthra commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -602,11 +724,15 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename)
             av_freep(&tf_model);
             av_freep(&model);
-
+	    printf("Failed to load\n");
```
Can either remove or use `av_log()`

apenugon replied:
Removed this

jailuthra commented on libavfilter/dnn/libtf_wrapper.h:
```diff
@@ -0,0 +1,683 @@
+/*
```
Is it possible to move this header into `compat/dnn` instead of `libavfilter/dnn`? Would be similar to where other shared library wrapper reside, example CUDA.

apenugon replied:
Moved to compat/dnn

jailuthra commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -31,9 +31,75 @@
+    MACRO(TF_DeleteSession)				    \
+    MACRO(TF_DeleteStatus)				    \
+    MACRO(TF_DeleteGraph)				    \
+    MACRO(TF_NewGraph)					    \
+    MACRO(TF_DeleteTensor)				    \
+    MACRO(TF_GraphOperationByName)			    \
+    MACRO(TF_OperationOutputType)			    \
+    MACRO(TF_NewStatus)					    \
+    MACRO(TF_GraphGetTensorShape)			    \
+    MACRO(TF_NewOperation)				    \
+    MACRO(TF_SetAttrType)				    \
+    MACRO(TF_SetAttrShape)				    \
+    MACRO(TF_FinishOperation)				    \
+    MACRO(TF_AllocateTensor)				    \
+    MACRO(TF_SetAttrTensor)				    \
+    MACRO(TF_AddInput)					    \
+    MACRO(TF_SetAttrInt)				    \
+    MACRO(TF_DataTypeSize)				    \
+    MACRO(TF_SetAttrString)				    \
+    MACRO(TF_SetAttrIntList)				    \
+    MACRO(TF_NewSessionOptions)				    \
+    MACRO(TF_SetConfig)					    \
+    MACRO(TF_DeleteSessionOptions)			    \
+    MACRO(TF_NewImportGraphDefOptions)			    \
+    MACRO(TF_ImportGraphDefOptionsSetDefaultDevice)	    \
+    MACRO(TF_GraphImportGraphDef)			    \
+    MACRO(TF_DeleteImportGraphDefOptions)		    \
+    MACRO(TF_DeleteBuffer)				    \
+    MACRO(TF_NewBuffer)					    \
+    MACRO(TF_Version)
+
+#define TF_LOAD_SYMBOL(sym) tf_model->tffns->sym = (PFN_##sym)TF_SYM_FUNC(tf_model->libtensorflow, #sym);
+
+#define PFN_DEF(name)  \
+    PFN_##name name;
+
+typedef struct TFFunctions {
+    FN_LIST(PFN_DEF)
+} TFFunctions;
+
```
Better to move this whole section to the wrapper header I think.

apenugon replied:
Moved to wrapper header

jailuthra commented on configure:
```diff
@@ -6316,7 +6317,6 @@ enabled libsoxr           && require libsoxr soxr.h soxr_create -lsoxr
 enabled libssh            && require_pkg_config libssh libssh libssh/sftp.h sftp_init
 enabled libspeex          && require_pkg_config libspeex speex speex/speex.h speex_decoder_init
 enabled libsrt            && require_pkg_config libsrt "srt >= 1.3.0" srt/srt.h srt_socket
-enabled libtensorflow     && require libtensorflow tensorflow/c/c_api.h TF_Version -ltensorflow
```
given that we still depend on the header files while building - would keeping it around like this work?

`enabled libtensorflow     && require_headers "tensorflow/c/c_api.h dlfcn.h"`

it's done similarly for frei0r filter

apenugon replied:
Added the tensorflow/c/c_api.h requirement but did not add the dlfcn.h requirement, since that only exists on Linux and the requirement for libdl or LoadLibrary is taken care of above anyway.

I'm not sure how the libfrei0r configuration works with Windows, since mingw32 won't find dlfcn.h.

jailuthra commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -665,26 +791,29 @@ void ff_dnn_free_model_tf(DNNModel **model)
         }
+
+	av_freep(&tf_model->tffns);
+	TF_FREE_FUNC(tf_model->libtensorflow);		
```
nit: indent mismatch here and a couple other places too.

apenugon replied:
All indents should be good now.

apenugon commented:
Addressed comments! Also I hadn't tested that case where tensorflow is not installed but a tensorflow filter is applied anyway - I also got a segfault when doing that with the super-resolution filter, and fixed that in the commit I just uploaded (it was an issue with the error type I was returning).

jailuthra committed:
Fix copyright

jailuthra reviewed (approved):
LGTM :ship:
